### DISCUSSION TITLE
iMessage Size Discrepancy

### DISCUSSION THREAD

**Discussion by tonyflo** [ReagentX/imessage-exporter#399] [Q&A] [answered]
I read through the previous discussions on this topic, but it's still not clear to me why the size that iMessage reports is different than imessage-exporter.

iCloud says my Messages take up 59.6 GB.
<img width="300" alt="Screenshot 2024-12-05 at 2 32 53 PM" src="https://github.com/user-attachments/assets/ee4067ff-d3df-41a4-9966-4be7963988c0">

My filesystem closely reflects that size at 59.92 GB.
<img width="450" alt="Screenshot 2024-12-05 at 2 32 08 PM" src="https://github.com/user-attachments/assets/efa41fe6-8765-4f05-adbf-17361da90e83">

As expected, most of that size is from Attachments.
<img width="350" alt="Screenshot 2024-12-05 at 2 32 08 PM" src="https://github.com/user-attachments/assets/ad203cba-736c-4e6f-8987-caf1e1558ae0">

However, diagnostics only detects around 30 GB of attachments.
<img width="450" alt="Screenshot 2024-12-05 at 2 32 08 PM" src="https://github.com/user-attachments/assets/f31a7ab5-e979-429c-ab99-95886103e91d">

**Comment by ReagentX** [comment, marked as answer]
Attachment diagnostic data is explained [here](https://github.com/ReagentX/imessage-exporter/blob/develop/docs/diagnostics.md#attachment-diagnostic-data).

**Reply by tonyflo** [reply]
Thank you for linking the diagnostic data explanation. You mention in there:

> The sum of the total_bytes column in the attachments table. I don't know why they are different, but the former is the actual storage taken up by iMessage attachments.

The actual storage taken up by the Attachments folder is 51.49 GB, not 32.89 GB as reported by diagnostics.

Based on the diagnostics and export size, I am not confident that this tool is exporting the entire iMessage database due to the mismatch in what I seeing on the filesystem (~60 GB) and the size of the export (~30 GB).

**Reply by ReagentX** [reply]
`imessage-exporter` only looks at attachments actually referenced by the `attachment` table, it doesn't look at the disk directly–if there are files not referenced in the table, they cannot be exported in the first place. You can read the logic [here](https://github.com/ReagentX/imessage-exporter/blob/e1c7fa5e7066772b05c9e13f18e07430192215b0/imessage-database/src/tables/attachment.rs#L348-L430).

**Comment by tonyflo** [comment]
That's very helpful. I just learned a lot about the filesystem storage of the iMessage database.

What I did was fire up sqlite3:
`sqlite3 chat.db`

And listed all the attachments in my database:
`sqlite> SELECT filename FROM ATTACHMENT;`

I picked a random one:
> ~/Library/Messages/Attachments/76/06/5F4D844E-F2BA-47E9-A938-E0F719401D71/IMG_6432.jpeg

And looked in that directory with Finder:
![Screenshot 2024-12-05 at 3 24 44 PM](https://github.com/user-attachments/assets/608afd45-b9d3-4598-8936-c1100296aea3)

As you can see there is a JPEG (the still image), MOV (the live photo) and another file format called PVT. I found out that the PVT file contains both the live photo and still image and just so happens that the file sizes equal each other: JPEP + MOV = PVT (2.2 MB + 4.5 MB = 6.7 MB)

Now I guess because the chat.db is only referencing IMG_6432.jpeg then MOV and PVT files are not considered in the export. Is that true, and if so I think this could explain why iCloud and my filesystem are approximately 2x larger than the export size?

**Reply by ReagentX** [reply]
It's possible, but that may not be the only reason.
